host-apis/wasi-0.2.0/host_api.cpp

```diff
@@ -5,9 +5,18 @@
 size_t api::AsyncTask::select(std::vector<AsyncTask *> &tasks) {
   auto count = tasks.size();
   vector<WASIHandle<host_api::Pollable>::Borrowed> handles;
-  for (const auto task : tasks) {
-    handles.emplace_back(task->id());
+
+  for (size_t idx = 0; idx < count; ++idx) {
+    auto *task = tasks.at(idx);
+    auto id = task->id();
+
+    if (id == IMMEDIATE_TASK_HANDLE) {
+      return idx;
+    } else {
+      handles.emplace_back(id);
+    }
```

Comment:
While this is the obvious way of supporting internal tasks, unfortunately this way of integrating internal tasks means that they will always be run as much as possible, starving other tasks that might also be ready. E.g. if you're processing a large `Blob` while concurrently doing a `fetch`, the response to the `fetch` will only be processed once the entire `Blob` has been processed, instead of whenever it's actually available.

I think there's a decent way of handling this, though: if we encounter an `IMMEDIATE_TASK_HANDLE` in this loop, we can check if we have any other entries in `handles` already. If so, we can add an additional pollable that'll always be ready immediately, and do a poll of the entries before the `IMMEDIATE_TASK_HANDLE`. If any of those are ready, return the first one, otherwise return the `IMMEDIATE_TASK_HANDLE`.

I haven't tested this, but I think if you replace the entire `api::AsyncTask::select` method with the following code, things should either work, or be very close to working:

```cpp
static optional<wasi_clocks_monotonic_clock_own_pollable_t> immediately_ready;

size_t poll_handles(vector<WASIHandle<host_api::Pollable>::Borrowed> handles) {
  auto list = list_borrow_pollable_t{handles.data(), handles.size()};
  bindings_list_u32_t result{nullptr, 0};
  wasi_io_poll_poll(&list, &result);
  MOZ_ASSERT(result.len > 0);
  const auto ready_index = result.ptr[0];
  free(result.ptr);
  return ready_index;
}

size_t api::AsyncTask::select(std::vector<AsyncTask *> &tasks) {
  auto count = tasks.size();
  vector<WASIHandle<host_api::Pollable>::Borrowed> handles;
  for (size_t idx = 0; idx < count; ++idx) {
    auto *task = tasks.at(idx);
    auto id = task->id();

    if (id == IMMEDIATE_TASK_HANDLE) {
      if (handles.size() > 0) {
        if (!immediately_ready) {
          immediately_ready = wasi_clocks_monotonic_clock_subscribe_duration(0);
        }
        handles.emplace_back(immediately_ready.value().__handle);
        size_t ready_index = poll_handles(std::move(handles));
        if (ready_index <= idx) {
          return ready_index;
        }
      }
      return idx;
    }
    handles.emplace_back(id);
  }

  return poll_handles(std::move(handles));
}
```